contrib/ssh-copy-id

```diff
@@ -295,9 +299,25 @@ case "$REMOTE_VERSION" in
     populate_new_ids 0
     # in ssh below - to defend against quirky remote shells: use 'exec sh -c' to get POSIX;
     #     'cd' to be at $HOME; add a newline if it's missing; and all on one line, because tcsh.
-    [ "$DRY_RUN" ] || printf '%s\n' "$NEW_IDS" | \
+    if ! [ "$DRY_RUN" ] && ! [ "$USE_SFTP" ]; then
+      printf '%s\n' "$NEW_IDS" | \
       ssh "$@" "exec sh -c 'cd ; umask 077 ; mkdir -p .ssh && { [ -z "'`tail -1c .ssh/authorized_keys 2>/dev/null`'" ] || echo >> .ssh/authorized_keys ; } && cat >> .ssh/authorized_keys || exit 1 ; if type restorecon >/dev/null 2>&1 ; then restorecon -F .ssh .ssh/authorized_keys ; fi'" \
       || exit 1
+    fi
+    if ! [ "$DRY_RUN" ] && [ "$USE_SFTP" ] ; then
+      SHARED_CON_TMPFILE=$(mktemp ssh-copy-id-shared-connection.XXXXXXXXXXXX)
+      KEYS_TMPFILE=$(mktemp authorized_keys.XXXXXXXXXXXX)
+      ssh -f -N -M -S $SHARED_CON_TMPFILE "$@"
+      echo "mkdir .ssh" | sftp -b - -o "ControlPath=${SHARED_CON_TMPFILE}" "notnecessary" 2>/dev/null
+      echo "touch .ssh/authorized_keys" | sftp -b - -o "ControlPath=${SHARED_CON_TMPFILE}" "notnecessary"
+      echo "get .ssh/authorized_keys $KEYS_TMPFILE" | sftp -b - -o "ControlPath=$SHARED_CON_TMPFILE" "notnecessary" || exit 1
+      # add a newline if it's missing
+      [ -z "$(tail -1c $KEYS_TMPFILE 2>/dev/null)" ] || echo >> $KEYS_TMPFILE
+      printf '%s\n' "$NEW_IDS" >> $KEYS_TMPFILE
+      echo "put $KEYS_TMPFILE .ssh/authorized_keys" | sftp -b - -o "ControlPath=$SHARED_CON_TMPFILE" "notnecessary" || exit 1
+      ssh -S $SHARED_CON_TMPFILE -O exit "notnecessary"
+      trap "rm -f $KEYS_TMPFILE" EXIT
```

Comment:
this line directly after mktemp. Otherwise it's not executed if something fails.